docs/proposals/010-new-diff-type-system.md

```diff
@@ -0,0 +1,283 @@
+{
+	// A backward derivative method is currently responsible for 'setting' the differential values.
+	dpa = DifferentialPair<float>(dpa.p, d_out);
+	dpb = DifferentialPair<float>(dpb.p, d_out);
+}
+```
+
+Unfortunately, this makes little sense if we decide to use buffer or pointer types:
+```csharp
+struct DiffPtr<T> : IDifferentiable
+{
+	StructuredBuffer<T> bufferRef;
+	uint64 offset;
+	
+	[Differentiable] T get() { ... }
+	[Differentiable] void set(T t) { ... }
+	/* 
+	  Problem 1:
+	  We use custom derivatives for get() and set() to backprop and 
+	  read gradients. If DiffPtr<T> is differentiable, then get() and 
+	  set() need to operate on the *pair* type and not this struct type.
+	  There is no proper way to do this currently.
+	*/
+};
+
+[Differentiable]
+void add(DiffPtr<float> a, DiffPtr<float> b, DiffPtr<float> output)
+{
+	output.set(a.get() + b.get());
+}
+
+// Synthesized derivative:
+[Differentiable]
+void s_bwd_add(
+	inout DifferentialPair<DiffPtr<float>> a,
+	inout DifferentialPair<DiffPtr<float>> b,
+	inout DifferentialPair<DiffPtr<float>> output)
+{
+	/* 
+	   Problem 2:
+	   
+	   Current backward mode semantics require that the method assume that the differentials
+	   a.d and b.d are empty/zero, and it is the backward method's job to populate the result.
+	   
+	   It doesn't make sense to 'set' the differential part since it is a buffer ref.
+	   Rather, we want the user to provide the differential pointer, and use custom derivatives of 
+	   the getters/setters to propagate derivatives.
+	   
+	   This also means methods like dzero(), dadd() and dmul() make no sense 
+	   in the context of pointer types. They cannot be initialized within a derivative method.
+	*/
+}
+
+```
+
+## Workarounds
+At the moment the primary workaround is to use a **non-differentiable buffer type** with differentiable methods, and always initialize the object with two pointers for both the primal and differential buffers. This is how our `DiffTensorView<T>` object works. 
+Unfortunately, this is a rather hacky workaround with several drawbacks: 
+1.  `DiffTensorView<T>` does not conform to `IDifferentiable`, but is used for derivatives. This makes our type system less useful as checks for `is_subtype` from applications using reflection need workarounds to account for corner cases like these.
+2.  `DiffTensorView<T>` always has two buffer pointers even when used in non-differentiable methods. This is extra data in the struct, and potentially extra tensor allocations (we explicitly handle this case in `slangtorch` by leaving the diff part uninitialized if a primal method is invoked)
+3.  Higher-order derivatives don't work well with this workaround. Differentiating a method twice needs a set of 4 pointers, but we need to account for this ahead of time by using new types like `DiffDiffTensorView` that worsens the problem of carrying around extra data where its not required.
```

Comment:
How does higher order diff work with the new system?

Reply:
In this scenario, we can have a `DTensorView<T> : IDifferentiablePtrType` which can have a single buffer in primal code. 
It becomes `DifferentialPair<DTensorView<T>>` under the first derivative pass (2 buffers), and then becomes `DifferentialPair<DifferentialPair<<DTensorView<T>>>` under the second derivative pass (4 buffers).